# Step 1: Data Collection

In [34]:
# Import libraries
import os
import pandas as pd
import numpy as np

In [35]:
# Print current directory
print(os.getcwd())

/Users/phongporter/Documents/GITHUB/COS40007/week 3/ampc_portfolio


In [36]:
os.chdir("/Users/phongporter/Documents/GITHUB/COS40007/week 3/ampc_portfolio")

In [37]:
# Load the dataset from ampc2 (ampc_portfolio)
df_boning = pd.read_csv('Boning.csv')
df_slicing = pd.read_csv('Slicing.csv')

In [38]:
# Extract columns
extract_columns = ['Frame', 'Right Upper Arm x', 'Right Upper Arm y', 'Right Upper Arm z', 
                      'Left Upper Arm x', 'Left Upper Arm y', 'Left Upper Arm z']
boning_extracted = df_boning[extract_columns]
slicing_extracted = df_slicing[extract_columns]

#Print DataFrame with extracted columns
print(boning_extracted)
print(slicing_extracted)



       Frame  Right Upper Arm x  Right Upper Arm y  Right Upper Arm z  \
0          0           0.559333           0.024451           0.523876   
1          1           0.364502           0.174362           0.578967   
2          2          -0.041012           0.134000           0.285496   
3          3           0.007940           0.223349           0.133206   
4          4           0.418177           0.374025           0.080194   
...      ...                ...                ...                ...   
54175  54175          -0.859456          -0.295704          -2.242624   
54176  54176           0.154228           0.177665          -2.178331   
54177  54177           0.121470          -0.017822          -1.919329   
54178  54178          -0.145355           0.002809          -1.044747   
54179  54179          -0.177527           0.383617          -0.046803   

       Left Upper Arm x  Left Upper Arm y  Left Upper Arm z  
0              0.005296         -0.301528         -0.231850  

In [39]:
boning_extracted['class'] = 0
slicing_extracted['class'] = 1

# Merge the two datasets
df_combined_data = pd.concat([boning_extracted, slicing_extracted], ignore_index=True)

# Print the combined dataset
print(df_combined_data)

       Frame  Right Upper Arm x  Right Upper Arm y  Right Upper Arm z  \
0          0           0.559333           0.024451           0.523876   
1          1           0.364502           0.174362           0.578967   
2          2          -0.041012           0.134000           0.285496   
3          3           0.007940           0.223349           0.133206   
4          4           0.418177           0.374025           0.080194   
...      ...                ...                ...                ...   
72055  17875          -0.114328           0.439828          -0.081698   
72056  17876          -0.370897           0.463411          -0.037791   
72057  17877          -0.495828           0.108677          -0.183645   
72058  17878          -0.473204          -0.212558          -0.433903   
72059  17879          -0.337285          -0.520867          -0.449577   

       Left Upper Arm x  Left Upper Arm y  Left Upper Arm z  class  
0              0.005296         -0.301528         -0.2

/var/folders/c6/_p5br7tn1xj6tjc5l7wfj5nw0000gn/T/ipykernel_12589/2089278070.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  boning_extracted['class'] = 0
/var/folders/c6/_p5br7tn1xj6tjc5l7wfj5nw0000gn/T/ipykernel_12589/2089278070.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  slicing_extracted['class'] = 1


In [40]:
os.chdir("/Users/phongporter/Documents/GITHUB/COS40007/week 3/dataset_portfolio")

In [41]:
# Export to new file csv file
df_combined_data.to_csv('combined_data.csv', index=False)

# Step 2: Create Composite columns

Create some composite data points for your data as follows
1) Root mean square value of x and y
2) Root mean square value of y and z
3) Root mean square value of z and x
4) Root mean square value of x, y and z
5) Roll `(180 * atan2(accelY, sqrt(accelX * accelX + accelZ * accelZ))/PI)` 
6) Pitch `(180 * atan2(accelX, sqrt(accelY * accelY + accelZ * accelZ))/PI)`

So, you will get 12 more columns for 2 settings from the above. Merge this with your data in Step 1. Now you will have data file with 20 columns.

In [42]:
# Load combined data
df_combined_data = pd.read_csv('combined_data.csv')

## Column 8-13: 6 computed columns in step 2 for column Set 1

In [44]:
# 1 Root mean square value of x and y
df_combined_data['rms_xy_set_1'] = np.sqrt(np.mean(df_combined_data['Right Upper Arm x']**2)) + np.sqrt(np.mean(df_combined_data['Right Upper Arm y']**2))
# 2 Root mean square value of y and z
df_combined_data['rms_yz_set_1'] = np.sqrt(np.mean(df_combined_data['Right Upper Arm z']**2)) + np.sqrt(np.mean(df_combined_data['Right Upper Arm y']**2))
# 3 Root mean square value of x and z
df_combined_data['rms_zx_set_1'] = np.sqrt(np.mean(df_combined_data['Right Upper Arm z']**2)) + np.sqrt(np.mean(df_combined_data['Right Upper Arm x']**2))
# Root mean square value of x, y and z
df_combined_data['rms_xyz_set_1'] = np.sqrt(np.mean(df_combined_data['Right Upper Arm x']**2)) + np.sqrt(np.mean(df_combined_data['Right Upper Arm y']**2)) + np.sqrt(np.mean(df_combined_data['Right Upper Arm z']**2))

# Roll
df_combined_data['rms_roll_set_1'] = 180 * np.arctan2(df_combined_data['Right Upper Arm y'], np.sqrt(df_combined_data['Right Upper Arm x']**2 + df_combined_data['Right Upper Arm z']**2)) / np.pi
# Pitch
df_combined_data['rms_pitch_set_1'] = 180 * np.arctan2(df_combined_data['Right Upper Arm x'], np.sqrt(df_combined_data['Right Upper Arm y']**2 + df_combined_data['Right Upper Arm z']**2)) / np.pi


## Column 14-19: 6 computed columns in step 2 for column Set 2

In [45]:
# 1 Root mean square value of x and y
df_combined_data['rms_xy_set_2'] = np.sqrt(np.mean(df_combined_data['Left Upper Arm x']**2)) + np.sqrt(np.mean(df_combined_data['Left Upper Arm y']**2))
# 2 Root mean square value of y and z
df_combined_data['rms_yz_set_2'] = np.sqrt(np.mean(df_combined_data['Left Upper Arm z']**2)) + np.sqrt(np.mean(df_combined_data['Left Upper Arm y']**2))
# 3 Root mean square value of x and z
df_combined_data['rms_zx_set_2'] = np.sqrt(np.mean(df_combined_data['Left Upper Arm z']**2)) + np.sqrt(np.mean(df_combined_data['Left Upper Arm x']**2))
# Root mean square value of x, y and z
df_combined_data['rms_xyz_set_2'] = np.sqrt(np.mean(df_combined_data['Left Upper Arm x']**2)) + np.sqrt(np.mean(df_combined_data['Left Upper Arm y']**2)) + np.sqrt(np.mean(df_combined_data['Left Upper Arm z']**2))

# Roll
df_combined_data['rms_roll_set_2'] = 180 * np.arctan2(df_combined_data['Left Upper Arm y'], np.sqrt(df_combined_data['Left Upper Arm x']**2 + df_combined_data['Left Upper Arm z']**2)) / np.pi
# Pitch
df_combined_data['rms_pitch_set_2'] = 180 * np.arctan2(df_combined_data['Left Upper Arm x'], np.sqrt(df_combined_data['Left Upper Arm y']**2 + df_combined_data['Left Upper Arm z']**2)) / np.pi


In [50]:
updated_columns_step_2 = ['Frame', 
                          'rms_xy_set_1', 'rms_yz_set_1', 'rms_zx_set_1', 'rms_xyz_set_1', 'rms_roll_set_1', 'rms_pitch_set_1', 
                          'rms_xy_set_2', 'rms_yz_set_2', 'rms_zx_set_2', 'rms_xyz_set_2', 'rms_roll_set_2', 'rms_pitch_set_2', 
                          'class']
updated_data = df_combined_data[updated_columns_step_2]

In [48]:
# Print current directory
print(os.getcwd())

/Users/phongporter/Documents/GITHUB/COS40007/week 3/dataset_portfolio


In [51]:
# Save into new csv file
updated_data.to_csv('updated_combined_data.csv', index=False)

# Step 3: Data pre-processing and Feature computation

Now you will need to write code to create statistical features per minute (1 minute = consecutive 60 frames) from the 18 columns. Like we use in Studio 3 compute the following features for each column 2-19

In [96]:
from scipy.signal import find_peaks
from scipy.integrate import simpson as simps

In [55]:
# Read the updated data
df_updated_data = pd.read_csv('updated_combined_data.csv')

In [102]:
# number of frame per second
frame_per_mins = 60

# List all columns, except first and last column
columns_without_first_and_last = updated_data.columns[1:-1]

feature_rows = []

# Process each column in frame
for start in range(0, len(df_updated_data), frame_per_mins):
    if start + frame_per_mins > len(df_updated_data):
        break

    segment = df_updated_data.iloc[start:start + frame_per_mins]

    features = {}

    for column in columns_without_first_and_last:
        col_data = segment[column]

        features[column + '_mean'] = col_data.mean()
        features[column + '_std'] = col_data.std()
        features[column + '_min'] = col_data.min()
        features[column + '_max'] = col_data.max()

        # Calculate AUC
        features[column + '_auc'] = simps(col_data) # use last value

        # Calculate Peaks
        features[column + '_peaks'] = len(find_peaks(col_data)[0])

        # Append the class label
        features['class'] = segment['class'].values[0]
        
        # Append the features into the list
        feature_rows.append(features)

feature_df = pd.DataFrame(feature_rows)


# Save into new csv file
feature_df.to_csv('feature_computation_data.csv', index=False)
        


# Step 4: Training

## 1. Train-Test split (70/30)

In [106]:
from sklearn.model_selection import train_test_split

X = feature_df.drop('class', axis=1)
y = feature_df['class']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [127]:
from sklearn.metrics import accuracy_score

# Train SVM model
clf = svm.SVC()
clf.fit(X_train, y_train)

# Predict on the test set
y_predict = clf.predict(X_test)

# Calculate the accuracy
train_accuracy_score_step4 = accuracy_score(y_test, y_predict)

# Print the accuracy
print(f"Accuracy:  {train_accuracy_score_step4:.2f}")

Accuracy:  0.80


## 2. 10-fold cross validation

In [107]:
from sklearn import svm
from sklearn.model_selection import cross_val_score

clf = svm.SVC()
scores_10_fold_cross_validation = cross_val_score(clf, X, y, cv=10)
# print(scores)
print(f'Cross-validation scores: {scores_10_fold_cross_validation}')

# Median of the scores'
print(f'Median of the cross-validation scores: {scores_10_fold_cross_validation.mean():.2f}')

Cross-validation scores: [0.76837725 0.73509015 0.79389313 0.74392783 0.74392783 0.77723803
 0.75156142 0.76266482 0.78209577 0.76821652]
Median of the cross-validation scores: 0.76


## 3. 1 and 2 with hyper parameter tuning

In [108]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score

# define parameter range
param_grid = {'C': [0.1, 1, 10, 100, 1000],
              'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
              'kernel': ['rbf']}

grid = GridSearchCV(clf, param_grid, refit=True, verbose=3)

# fitting the model for grid search
grid.fit(X_train, y_train)

# Predict and measure accuracy
y_pred = grid.predict(X_test)
accuracy_train_ts_1_and_2= accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy_train_ts_1_and_2}')

cross_val_1_and_2 = cross_val_score(grid, X, y, cv=10)
median_cross_val_1_and_2 = cross_val_1_and_2.mean()
# print(scores)
print(f'Cross-validation scores: {median_cross_val_1_and_2:.2f}')

Fitting 5 folds for each of 25 candidates, totalling 125 fits
[CV 1/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.802 total time=   2.0s
[CV 2/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.803 total time=   2.0s
[CV 3/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.810 total time=   2.0s
[CV 4/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.811 total time=   2.1s
[CV 5/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.806 total time=   2.1s
[CV 1/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.802 total time=   2.0s
[CV 2/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.803 total time=   2.0s
[CV 3/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.810 total time=   2.0s
[CV 4/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.811 total time=   2.1s
[CV 5/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.806 total time=   2.0s
[CV 1/5] END .....C=0.1, gamma=0.01, kernel=rbf;, score=0.802 total time=   2.0s
[CV 2/5] END .....C=0.1, gamma=0.01, kernel=rbf

## 4. 1 and 2 with hyper parameter tuning and 10 best features

In [109]:
from sklearn.datasets import load_iris
from sklearn.feature_selection import SelectKBest, f_classif

# Select the top 100 features
X, y = load_iris(return_X_y=True)
X.shape

(150, 4)

In [111]:
k = min(100, X.shape[1]) # Adject number from 100 to minimum number of features
k_best = SelectKBest(f_classif, k=k)
X_k_best = k_best.fit_transform(X, y)
X_k_best.shape

(150, 4)

In [119]:
X_k_train, X_k_test, y_k_train, y_k_test = train_test_split(X_k_best, y, test_size=0.3, random_state=1)


In [123]:
# Train the Support Vector Classifier without Hyper-parameter Tuning
clf_k_best = svm.SVC()
clf_k_best.fit(X_k_train, y_k_train)

# Print predict results
y_pred_k_best = clf_k_best.predict(X_k_test)
train_test_accuracy_kbest_1_and_2 = accuracy_score(y_k_test, y_pred_k_best)
print(f'Accuracy with K-best: {train_test_accuracy_kbest_1_and_2}')

Accuracy with K-best: 0.9777777777777777


In [128]:
scores_kbest_1_and_2 = cross_val_score(clf_k_best, X_k_best, y, cv=10)

# Print cross-validation scores
print(f'Cross-validation scores: {scores_kbest_1_and_2}')
print(f'Median of the cross-validation scores: {scores_kbest_1_and_2.mean():.2f}')

Cross-validation scores: [1.         0.93333333 1.         1.         1.         0.93333333
 0.93333333 0.93333333 1.         1.        ]
Median of the cross-validation scores: 0.97


## 5. 1 and 2 with hyper parameter tuning and 10 principal components

In [124]:
from sklearn.decomposition import PCA
X_pca = PCA(n_components=4).fit_transform(X)

In [125]:
X_pca_train, X_pca_test, y_pca_train, y_pca_test = train_test_split(X_pca, y, test_size=0.3, random_state=1)

clf_pca = svm.SVC()
clf_pca.fit(X_pca_train, y_pca_train)

y_pca_pred = clf_pca.predict(X_pca_test)
train_test_accuracy_pca_1_and_2 = accuracy_score(y_pca_test, y_pca_pred)
print(f'Accuracy with PCA: {train_test_accuracy_pca_1_and_2:.2f}')

Accuracy with PCA: 0.98


In [ ]:
# Train the Support Vector Classifier without Hyper-parameter Tuning
clf_k_best = svm.SVC()
clf_k_best.fit(X_k_train, y_k_train)

# Print predict results
y_pred_k_best = clf_k_best.predict(X_k_test)
train_test_accuracy_kbest_1_and_2 = accuracy_score(y_k_test, y_pred_k_best)
print(f'Accuracy with K-best: {train_test_accuracy_kbest_1_and_2}')

Accuracy with K-best: 0.9777777777777777


In [126]:
scores_pca_1_and_2 = cross_val_score(clf_pca, X_pca, y, cv=10)
print(f'Cross-validation scores: {scores_pca_1_and_2.mean():.2f}')

Cross-validation scores: 0.97


## Create the summary table of your outcome similar to Studio 3 Activity 6.

In [129]:
import pandas as pd

summary_data = {
    'SVM Model': ['Original features', 'With hyperparameter tuning', 'With feature selection and hyperparameter tuning', 'With PCA and hyperparameter tuning'],
    'Train-test split': [f'{train_accuracy_score_step4}%', f'{accuracy_train_ts_1_and_2}%', f'{train_test_accuracy_kbest_1_and_2}%', f'{train_test_accuracy_pca_1_and_2}%'],
    'Cross validation': [f'{scores_10_fold_cross_validation.mean()}%', f'{median_cross_val_1_and_2}%', f'{scores_kbest_1_and_2.mean()}%', f'{scores_pca_1_and_2.mean()}%']
}

summary_table = pd.DataFrame(summary_data)
print(summary_table)

                                          SVM Model     Train-test split  \
0                                 Original features   0.797641073080481%   
1                        With hyperparameter tuning                 1.0%   
2  With feature selection and hyperparameter tuning  0.9777777777777777%   
3                With PCA and hyperparameter tuning  0.9777777777777777%   

      Cross validation  
0  0.7626992736012228%  
1  0.7593672909762732%  
2  0.9733333333333334%  
3  0.9733333333333334%  


## Now, train SGD, RandomForest and MLP classifier with original dataset similar to Studio 3 activity 7 and generate the outcome table

### Train SGD

In [130]:
# Print current directory
print(os.getcwd())

/Users/phongporter/Documents/GITHUB/COS40007/week 3/dataset_portfolio


In [131]:
from sklearn.linear_model import SGDClassifier

# df_all_data = pd.read_csv('all_data.csv')

# X = df_all_data.iloc[:, :-1]
# y = df_all_data.iloc[:, -1]

# X_train, X_test, y_train, y = train_test_split(X, y, test_size=0.3, random_state=1)

clf_SGD = SGDClassifier()
clf_SGD.fit(X_train, y_train)

y_pred_SGD = clf_SGD.predict(X_test)
train_test_accuracy_SGD = accuracy_score(y_test, y_pred_SGD)
print(f'Accuracy with SGD: {train_test_accuracy_SGD:.2f}')

scores_SGD = cross_val_score(clf_SGD, X, y, cv=10)
print(f'Cross-validation scores: {scores_SGD.mean():.2f}')

Accuracy with SGD: 0.79
Cross-validation scores: 0.78


### Train RandomForest

In [132]:

from sklearn.ensemble import RandomForestClassifier

clf_RF = RandomForestClassifier()
clf_RF.fit(X_train, y_train)

y_pred_RF = clf_RF.predict(X_test)
train_test_accuracy_RF = accuracy_score(y_test, y_pred_RF)
print(f'Accuracy with Random Forest: {train_test_accuracy_RF:.2f}')

scores_RF = cross_val_score(clf_RF, X, y, cv=10)
print(f'Cross-validation scores: {scores_RF.mean():.2f}')

Accuracy with Random Forest: 1.00
Cross-validation scores: 0.96


### Train MLP

In [133]:
from sklearn.neural_network import MLPClassifier

clf_MLP = MLPClassifier()
clf_MLP.fit(X_train, y_train)

y_pred_MLP = clf_MLP.predict(X_test)
train_test_accuracy_MLP_activity7 = accuracy_score(y_test, y_pred_MLP)
print(f'Accuracy with MLP: {train_test_accuracy_MLP_activity7:.2f}')

scores_MLP = cross_val_score(clf_MLP, X, y, cv=10)
print(f'Cross-validation scores: {scores_MLP.mean():.2f}')

Accuracy with MLP: 0.88


/opt/miniconda3/envs/env_c_w1/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/miniconda3/envs/env_c_w1/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/miniconda3/envs/env_c_w1/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/miniconda3/envs/env_c_w1/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/minicon

Cross-validation scores: 0.98


/opt/miniconda3/envs/env_c_w1/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


### Outcome table

In [134]:
import pandas as pd

summary_data = {
    'SVM Model': ['SVM', 'SGD', 'RandomForest', 'MLP'],
    'Train-test split': [f'{train_accuracy_score_step4}%', f'{train_test_accuracy_SGD}%', f'{train_test_accuracy_RF}%', f'{train_test_accuracy_MLP_activity7}%'],
    'Cross validation': [f'{scores_10_fold_cross_validation.mean()}%', f'{scores_SGD.mean()}%', f'{scores_RF.mean()}%', f'{scores_MLP.mean()}%']
}

summary_table = pd.DataFrame(summary_data)
print(summary_table)

      SVM Model     Train-test split     Cross validation
0           SVM   0.797641073080481%  0.7626992736012228%
1           SGD  0.7897779833487512%                0.78%
2  RandomForest                 1.0%                0.96%
3           MLP  0.8818223866790009%  0.9800000000000001%


| Model                                           | Train-test split | Cross validation |
|-----------------------------------------------------|------------------|------------------|
| SVM                                   | 80%              | 76%              |
| SGD                          | 79%              | 78%              |
| RandomForest    | 100%              | 96%              |
| MLP                  | 89%              | 98%              |

# Step 5: Model Selection

## 1) Which SVM model will be the best for your problem

feature selection/PCA and hyperparameter tuning

## 2) Which ML model will be the best for your problem

MLP